shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py

```diff
@@ -156,13 +156,64 @@ class ProtocolSchemaV6(BaseModel):
     robot: Robot
     pipettes: Dict[str, Pipette]
     labware: Dict[str, Labware]
-    labwareDefinitions: Dict[str, LabwareDefinition]
-    commands: List[Command]
     modules: Optional[Dict[str, Module]]
     liquids: Optional[Dict[str, Liquid]]
+    labwareDefinitions: Dict[str, LabwareDefinition]
+    commands: List[Command]
     commandAnnotations: Optional[List[CommandAnnotation]]
     designerApplication: Optional[DesignerApplication]
 
     class Config:
         # added for constructing the class with field name instead of alias
         allow_population_by_field_name = True
+
+    @validator("commands")
+    def _validate_commands(
+        cls,
+        value: List[Command],
+        values: Dict[str, Any],
+    ) -> List[Command]:
+        pipette_ids = set(values["pipettes"].keys()) if "pipettes" in values else set()
+        labware_ids = set(values["labware"].keys()) if "labware" in values else set()
+        module_ids = (
+            set(values["modules"].keys())
+            if "modules" in values and values["modules"]
+            else set()
+        )
+        liquid_ids = (
+            set(values["liquids"].keys())
+            if "liquids" in values and values["liquids"]
+            else set()
+        )
+
+        for command in value:
```

Comment:
```suggestion
        for index, command in enumerate(value):
```

See my below suggestion about changing the error messages to include the command's index, not just its type.